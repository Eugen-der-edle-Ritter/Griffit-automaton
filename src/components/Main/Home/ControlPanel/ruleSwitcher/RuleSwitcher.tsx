import React, { FC } from 'react';
import styled from '@emotion/styled';
import { useSelector, useDispatch } from 'react-redux';
import { setRule } from './ruleSlice';

import { Rule } from '@/modules/types';
import { RootState } from '@/app/store';

export interface RuleSwitcherProps {
  label: string;
  options: Rule[];
}

export const RuleSwitcher: FC<RuleSwitcherProps> = ({ label, options }) => {
  const dispatch = useDispatch();
  const isPlaying = useSelector((state: RootState) => state.play.value);
  return (
    <Div>
      <Label>{label}: </Label>
      <Select
        data-testid="select"
        onChange={(e) => dispatch(setRule(e.target.value))}
        disabled={isPlaying}
      >
        {options.map((option, index) => (
          <Option key={index} value={option}>
            {option}
          </Option>
        ))}
      </Select>
    </Div>
  );
};

const Div = styled.div``;
const Label = styled.label`
  margin-right: 0.5rem;
  color: #e6e6e6;
  font-size: 0.85rem;
`;
const Select = styled.select`
  height: 1.6rem;
  padding-left: 0.2rem;
  color: #dddddd;
  background-color: #383b40;
  border: 0.5px solid #010101;
  border-radius: 0.5rem;
  outline: none;
  cursor: pointer;
  &:disabled {
    cursor: not-allowed;
  }
`;
const Option = styled.option``;
