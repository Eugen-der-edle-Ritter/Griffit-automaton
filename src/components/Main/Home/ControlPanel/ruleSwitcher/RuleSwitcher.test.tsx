import React from 'react';
import { Provider, useDispatch } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';

import { store } from '@/automaton/automaton';

import { RuleSwitcher } from './RuleSwitcher';

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

describe('RuleSwitcher test cases', () => {
  it('RuleSwitcher changing value', () => {
    const mockDispatch = jest.fn();

    (useDispatch as jest.Mock).mockReturnValue(mockDispatch);

    const { asFragment } = render(
      <Provider store={store}>
        <RuleSwitcher label="test" options={['Hash', 'Demons', 'Venus']} />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();

    fireEvent.change(screen.getByTestId('select'), {
      target: { value: 'Demons' },
    });

    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });
});
