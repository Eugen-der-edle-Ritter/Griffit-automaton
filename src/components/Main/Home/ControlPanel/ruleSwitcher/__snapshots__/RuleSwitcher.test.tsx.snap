// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleSwitcher test cases RuleSwitcher changing value 1`] = `
<DocumentFragment>
  <div
    class="css-0"
  >
    <label
      class="css-jm2r67"
    >
      test: 
    </label>
    <select
      class="css-vqx920"
      data-testid="select"
    >
      <option
        class="css-0"
        value="Hash"
      >
        Hash
      </option>
      <option
        class="css-0"
        value="Demons"
      >
        Demons
      </option>
      <option
        class="css-0"
        value="Venus"
      >
        Venus
      </option>
    </select>
  </div>
</DocumentFragment>
`;
